client/question/services/ParentPageService.js

```diff
@@ -0,0 +1,53 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service that handles all TIE interactions with its parent
+ * page.
+ */
+tie.factory('ParentPageService', [
+  '$window', 'PARENT_PAGE_URL_ORIGIN',
+  function($window, PARENT_PAGE_URL_ORIGIN) {
+    /**
+     * Used to define the raw code message type.
+     * @type {String}
+     * @constant
+     */
+    var MESSAGE_TYPE_RAW_CODE = 'raw_code';
+
+    /**
+     * Sends the message to the parent page, if it exists, via postMessage.
+     *
+     * @param {String} messageType Type of message that is being sent.
+     * @param {String} messageData Data to be sent to parent page.
+     */
+    var sendMessage = function(messageType, messageData) {
+      if (PARENT_PAGE_URL_ORIGIN) {
```

Comment:
Actually wait hang on. Shouldn't this also check whether $window.parent exists?

Reply:
$window.parent will always exist since $window.parent is itself if no explicit parent is present. (See https://developer.mozilla.org/en-US/docs/Web/API/Window/parent). I'll be sure to add the checks in the next PR though regarding making sure that the $window.parent actually matches the PARENT_PAGE_URL_ORIGIN. Or is it now better to include it in this PR?

Reply:
Ah, OK. I think you might as well include it here, yep (and add tests as needed).